File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.size()%2 != 0) return false;
        
        stack<char> st;

        for(auto ch: s){
            if(ch == '(' || ch == '[' || ch == '{'){
                st.push(ch);
            }
            else{
                if(st.empty()) return false;
                char t = st.top();
                st.pop();
                if((ch == ')' && t != '(') || (ch == ']' && t != '[') || (ch == '}' && t != '{')) 
                    return false;
            }
        }
        return st.empty();
    }
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for st in s:
            if (st == '(' or st == '[' or st == '{'):
                stack.append(st)
            else:
                if not stack: return False
                top = stack.pop()
                if ((st == ')' and top != '(') or
                    (st == ']' and top != '[') or
                    (st == '}' and top != '{') ):
                    return False
        
        return not stack